src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/Razor/GenerateMethodCodeActionResolver.cs

```diff
@@ -0,0 +1,112 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.LanguageServer.CodeActions.Models;
+using Microsoft.AspNetCore.Razor.LanguageServer.Common;
+using Microsoft.AspNetCore.Razor.LanguageServer.Formatting;
+using Microsoft.AspNetCore.Razor.Utilities;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Newtonsoft.Json.Linq;
+using CSharpSyntaxFactory = Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.CodeActions.Razor;
+internal class GenerateMethodCodeActionResolver : IRazorCodeActionResolver
+{
+    public string Action => LanguageServerConstants.CodeActions.GenerateMethod;
+    private readonly DocumentContextFactory _documentContextFactory;
+    private readonly RazorLSPOptions _razorLSPOptions;
+    private static readonly string s_generatedCode = $$"""0private void $$MethodName$$(){{Environment.NewLine}}0{{{Environment.NewLine}}1throw new System.NotImplementedException();{{Environment.NewLine}}0}""";
+
+    public GenerateMethodCodeActionResolver(DocumentContextFactory documentContextFactory, RazorLSPOptionsMonitor razorLSPOptionsMonitor)
+    {
+        _documentContextFactory = documentContextFactory;
+        _razorLSPOptions = razorLSPOptionsMonitor.CurrentValue;
+    }
+
+    public async Task<WorkspaceEdit?> ResolveAsync(JObject data, CancellationToken cancellationToken)
+    {
+        if (data is null)
+        {
+            return null;
+        }
+
+        var actionParams = data.ToObject<GenerateMethodCodeActionParams>();
+        if (actionParams is null)
+        {
+            return null;
+        }
+
+        var documentContext = await _documentContextFactory.TryCreateForOpenDocumentAsync(actionParams.Uri, cancellationToken).ConfigureAwait(false);
+        if (documentContext is null)
+        {
+            return null;
+        }
+
+        var code = await documentContext.GetCodeDocumentAsync(cancellationToken).ConfigureAwait(false);
+        var uriPath = FilePathNormalizer.Normalize(actionParams.Uri.GetAbsoluteOrUNCPath());
+        var codeBehindPath = $"{uriPath}.cs";
+        if (!File.Exists(codeBehindPath))
+        {
+            return GenerateMethodInCodeBlock(code, actionParams);
+        }
+
+        var razorCSharpText = (await documentContext.GetCSharpSourceTextAsync(cancellationToken).ConfigureAwait(false)).ToString();
+        var razorNamespace = razorCSharpText[(razorCSharpText.IndexOf("namespace ") + 10)..razorCSharpText.IndexOf("\r\n{")];
+        var mock = CSharpSyntaxFactory.ParseCompilationUnit(File.ReadAllText(codeBehindPath));
```

Comment:
Unfortunately I am not seeing any overload we can use to avoid reading the entire file here since we do not have any information about the code behind file aside from a uri path at this point. 😞